feat: add multi select element

JennaXiao3 opened the pull request:
## Notion ticket link
<!-- Please replace with your ticket's URL -->
[Add multi-select component to Question Editor](https://www.notion.so/uwblueprintexecs/Add-multi-select-component-to-Question-Editor-125650fdb8be4b0db5cf66e95fd5b1c3?pvs=4)


<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->
## Implementation description
* created a multi-select folder with a few components (repeated from multiple choice) under modals
* created MultiSelectElement.jsx
* edited a few question-related components to incorporate multi-select modal


<!-- What should the reviewer do to verify your changes? Describe expected results and include screenshots when appropriate -->
## Steps to test
1. Go to [localhost](http://localhost:3000/admin/create-question)


<!-- Draw attention to the substantial parts of your PR or anything you'd like a second opinion on -->
## What should reviewers focus on?
* 


## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

JennaXiao3 committed:
initial commit

JennaXiao3 committed:
completed all

joyce-shi reviewed (changes requested):
looks great! i'm happy you got it working. 🤠  but instead of duplicating the code from the Multiple Choice logic, would you be able to reuse code where possible?

in particular, `MultiSelectModal`, `MultiSelectOption`, `SelectOptionCount`, `MultiSelectData` and `updatedMultiSelectOption` are almost exact duplicates of their Multiple Choice counterparts!

JennaXiao3 commented:
> looks great! i'm happy you got it working. 🤠 but instead of duplicating the code from the Multiple Choice logic, would you be able to reuse code where possible?
> 
> in particular, `MultiSelectModal`, `MultiSelectOption`, `SelectOptionCount`, `MultiSelectData` and `updatedMultiSelectOption` are almost exact duplicates of their Multiple Choice counterparts!

If I want to reuse it, should I rename the components?

joyce-shi commented:
> > looks great! i'm happy you got it working. 🤠 but instead of duplicating the code from the Multiple Choice logic, would you be able to reuse code where possible?
> > in particular, `MultiSelectModal`, `MultiSelectOption`, `SelectOptionCount`, `MultiSelectData` and `updatedMultiSelectOption` are almost exact duplicates of their Multiple Choice counterparts!
> 
> If I want to reuse it, should I rename the components?

use your best judgement! feel free to rename existing files, move things around, whatever you think makes for the cleanest code! (:

JennaXiao3 committed:
passed state 'isMultiSelect' in context instead

JennaXiao3 commented:
> > > looks great! i'm happy you got it working. 🤠 but instead of duplicating the code from the Multiple Choice logic, would you be able to reuse code where possible?
> > > in particular, `MultiSelectModal`, `MultiSelectOption`, `SelectOptionCount`, `MultiSelectData` and `updatedMultiSelectOption` are almost exact duplicates of their Multiple Choice counterparts!
> > 
> > 
> > If I want to reuse it, should I rename the components?
> 
> use your best judgement! feel free to rename existing files, move things around, whatever you think makes for the cleanest code! (:

I've added a state called "isMultiSelect" and added necessary props to some multiple choices components when needed. Is this okay? If so, I will change the names of the multiple choice components to "multichoice" or something to show that it's used by multiple choice and multi-select

joyce-shi commented on frontend/src/components/pages/admin/QuestionPage.tsx:
```diff
@@ -32,6 +33,8 @@ const QuestionPage = (): React.ReactElement => {
           setShowAddShortAnswerModal,
           showAddMultipleChoiceModal,
           setShowAddMultipleChoiceModal,
+          isMultiSelect,
+          setIsMultiSelect,
```
I think we can keep the naming consistent here!

```suggestion
          showAddMultiSelectModal,
          setShowAddMultiSelectModal,
```

joyce-shi replied:
I think we can change the naming here to be consistent with the other question elements!

```suggestion
          showAddMultiSelectModal,
          setShowAddMultiSelectModal,
```

joyce-shi commented on frontend/src/components/question-creation/QuestionSidebarItem.tsx:
```diff
@@ -48,6 +49,11 @@ const QuestionSidebarItem = ({
+            break;
+          case QuestionElementType.MULTI_SELECT:
+            setShowAddMultipleChoiceModal(true);
+            setIsMultiSelect(true);
```
I think we can just set one to true and if either are true, we can show the modal! i think you'll have to update `AddMultipleChoiceModal` where that logic is defined 🤔 
```suggestion
            setShowAddMultiSelectModal(true);
```

joyce-shi commented on frontend/src/components/question-creation/question-elements/MultipleChoiceElement.tsx:
```diff
@@ -18,33 +20,53 @@ import EditMultipleChoiceModal from "./modals/multiple-choice/EditMultipleChoice
+                {option.value}
+              </Radio>
+            ))}
+          </VStack>
```
I think **lines 39 - 51** and **lines 55 - 67** are identical, can refactor them into a common constant?

joyce-shi commented on frontend/src/components/question-creation/question-elements/MultipleChoiceElement.tsx:
```diff
@@ -18,33 +20,53 @@ import EditMultipleChoiceModal from "./modals/multiple-choice/EditMultipleChoice
 const MultipleChoiceElement = ({
   id,
   data,
+  isMultiSelect,
```
nit: instead of having a prop `isMultiSelect`, can this component have a `type` prop that is of type `QuestionElementType.MULTIPLE_CHOICE | QuestionElementType.MULTI_SELECT` instead?

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/multiple-choice/AddMultipleChoiceModal.tsx:
```diff
@@ -15,26 +15,32 @@ const AddMultipleChoiceModal = (): React.ReactElement => {
     onClose();
   };
 
+  const multiData = isMultiSelect
```
nit: `multiData` -> `type` and then i think you can use shorthand

JennaXiao3 replied:
Do you mean 
```type multiData = showAddMultiSelectModal ? QuestionElementType.MULTI_SELECT : QuestionElementType.MULTIPLE_CHOICE;```

This doesn't seem to work for some reason...

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx:
```diff
@@ -61,7 +64,7 @@ const MultipleChoiceModal = ({
     } else if (correctOptionCount === 0) {
       setNoCorrectOptionError(true);
-    } else if (correctOptionCount > 1) {
+    } else if (!isMultiSelect && correctOptionCount > 1) {
```
nit: instead of `!isMultiSelect` can we use `isMultipleChoice` here for more better readability?

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/multiple-choice/EditMultipleChoiceModal.tsx:
```diff
@@ -38,6 +40,7 @@ const EditMultipleChoiceModal = ({
+      isMultiSelect={isMultiSelect}
       isOpen={isOpen}
       onClose={closeModal}
       onConfirm={updateMultipleChoiceElement}
```
same as above! if we could rename this modal to `EditMultiDataModal.tsx` or something, that would be awesome! and the folder should probably be renamed to `multi-option` as well!

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx:
```diff
@@ -76,7 +79,11 @@ const MultipleChoiceModal = ({
+      }
     >
       <VStack spacing="10" width="100%">
         {noCorrectOptionError && (
```
same as above! `MultipleChoiceModal` -> `MultiOptionModal`!

joyce-shi commented on frontend/src/components/question-creation/QuestionElementItem.tsx:
```diff
@@ -39,12 +39,20 @@ const renderQuestionContent = (content: QuestionElement) => {
+      return (
+        <MultipleChoiceElement
+          key={id}
+          data={data as MultipleChoiceData}
```
would you be able to rename the type to something more general since we are using it for both cases? maybe something like `MultiOptionData`!

joyce-shi commented on frontend/src/components/question-creation/question-elements/MultipleChoiceElement.tsx:
```diff
@@ -59,6 +81,7 @@ const MultipleChoiceElement = ({
       <EditMultipleChoiceModal
         data={data}
         id={id}
+        isMultiSelect={isMultiSelect}
```
nit: `isMultiSelect ` ->`type`

joyce-shi reviewed (commented):
looks a lot better! left some thoughts (:

i think overall, i would prefer if we kept the logic to show multiple choice and multi select separate! i.e can have separate `setShowMultipleChoiceModal` and `setShowMultiSelectModal` values (but should not have both set to true at the same time). left some more detailed thoughts on this below!

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/multiple-choice/AddMultipleChoiceModal.tsx:
```diff
@@ -15,26 +15,32 @@ const AddMultipleChoiceModal = (): React.ReactElement => {
   return (
     <MultipleChoiceModal
+      isMultiSelect={isMultiSelect}
       isOpen={showAddMultipleChoiceModal}
```
I think instead you can do something like `isMultipleChoice ? showAddMultipleChoiceModal : showAddMultiSelectModal` from the context value!

JennaXiao3 committed:
addressed pr comments

joyce-shi commented on frontend/src/components/question-creation/question-elements/MultipleChoiceElement.tsx:
```diff
@@ -11,40 +13,52 @@ import {
   ] = React.useState(false);
+  const InputType =
+    type === QuestionElementType.MULTIPLE_CHOICE ? Checkbox : Radio;
+
```
is this constant used somewhere? 👀

JennaXiao3 replied:
WHOOPS

joyce-shi commented on frontend/src/components/question-creation/QuestionElementItem.tsx:
```diff
@@ -39,12 +39,20 @@ const renderQuestionContent = (content: QuestionElement) => {
+      return (
+        <MultipleChoiceElement
+          key={id}
+          data={data as MultipleChoiceData}
```
Could you update `MultipleChoiceData` -> `MultiOptionData` or something like that?

joyce-shi commented on frontend/src/components/question-creation/question-elements/MultipleChoiceElement.tsx:
```diff
@@ -11,40 +13,52 @@ import {
+  type: QuestionElementType;
 }
 
 const MultipleChoiceElement = ({
```
Could you update `MultipleChoiceElement` -> `MultiOptionElement` or something like that?

joyce-shi commented on frontend/src/components/question-creation/question-elements/modals/multiple-choice/AddMultipleChoiceModal.tsx:
```diff
@@ -15,29 +15,41 @@ const AddMultipleChoiceModal = (): React.ReactElement => {
     setShowAddMultipleChoiceModal(false);
+    setShowAddMultiSelectModal(false);
     onClose();
   };
```
Could you update `AddMultipleChoiceModal` -> `AddMultiOptionModal` or something like that? also all the references to **Multiple Choice** in this file! & same for the `EditMultipleChoiceModal` and `MultipleChoiceModal`. You can also update the folder naming!

joyce-shi reviewed (changes requested):
looks so good! just left a couple comments about renaming the components that are used by both **Multiple Choice** and **Multi-Select** to something general. after that, should be good to go!

JennaXiao3 committed:
resolved pr comments

joyce-shi reviewed (approved):
LGTM

JennaXiao3 committed:
resolved some conflicts

JennaXiao3 committed:
Merge remote-tracking branch 'origin/staging' into jenna/multi-select
